test/cpp/tensorexpr/test_expr.cpp

```diff
@@ -556,6 +556,65 @@ TEST(Expr, DynamicShapeAdd) {
   testWithSize(37);
 }
 
+TEST(Expr, OutOfBounds) {
+  ExprHandle N(10);
+  ExprHandle start(0);
+  ExprHandle stop(15);
+  VarHandle i("i", kInt);
+
+  BufHandle X("X", {N}, kInt);
+
+  auto body = Store::make(X, {i}, i);
+  auto stmt = For::make(i, start, stop, body);
+
+  PaddedBuffer<int> data(20);
+
+  EXPECT_ANY_THROW(SimpleIREvaluator(stmt, {X})(data));
+}
+
+TEST(Expr, OutOfBounds2d) {
+  std::vector<std::pair<int, int>> size_options = {{10, 15}, {15, 10}};
+  for (auto sizes : size_options) {
+    ExprHandle N(sizes.first);
+    ExprHandle M(sizes.second);
+    ExprHandle start(0);
+    ExprHandle stopInner(15);
+    ExprHandle stopOuter(15);
+    VarHandle i("i", kInt);
+    VarHandle j("j", kInt);
+
+    BufHandle X("X", {N, M}, kInt);
+
+    auto body = Store::make(X, {i, j}, i);
+    auto inner = For::make(j, start, stopInner, body);
```

Comment:
Nit: we don't need to create ExprHandle for immediates, I think the default conversion would work as well, so we can write this as:
```cpp
auto inner = For::make(j, 0, 15, body);
```

Similarly, when we declare `X` we can write:
```cpp
BufHandle X("X", {sizes.first, sizes.second}, kInt);
```

Generally, everything with "Handle" is supposed to be user-friendly and have operators overloaded (e.g. we can write `i + 5` to generate an `Add` expression handle), conversions from default types defined, etc. It is when we work with expr pointers we have to be explicit in everything we do.